src/analysis/normed_space/normed_group_hom_completion.lean

```diff
@@ -0,0 +1,271 @@
+* `normed_group.to_compl : normed_group_hom G (completion G)`: the canonical map from `G` to its
+  completion, as a normed group hom
+* `normed_group_hom.completion_to_compl`: the above diagram indeed commutes.
+* `normed_group_hom.norm_completion_le`: `∥f.completion∥ ≤ ∥f∥`
+* `normed_group_hom.ker_le_ker_completion`: the kernel of `f.completion` contains the image of the
+  kernel of `f`.
+* `normed_group_hom.ker_completion`: the kernel of `f.completion` is the closure of the image of the
+  kernel of `f` under an assumption that `f` is quantitatively surjective onto its image.
+-/
+
+noncomputable theory
+
+open set normed_group_hom uniform_space uniform_space.completion
+
+variables {G : Type*} [semi_normed_group G]
+variables {H : Type*} [semi_normed_group H]
+variables {K : Type*} [semi_normed_group K]
+
+/-- The normed group hom induced between completions. -/
+def normed_group_hom.completion (f : normed_group_hom G H) :
+  normed_group_hom (completion G) (completion H) :=
+{ to_fun := completion.map f,
+  map_add' := begin
+    intros x y,
+    apply induction_on₂ x y,
+    apply is_closed_eq,
+    exact continuous_map.comp continuous_add,
+    exact (continuous_map.comp continuous_fst).add (continuous_map.comp continuous_snd),
+    intros a b,
+    norm_cast,
+    simp [map_coe f.uniform_continuous],
+    norm_cast
+  end,
+  bound' := begin
+    use ∥f∥,
+    intro y,
+    apply completion.induction_on y,
+    { exact is_closed_le (continuous_norm.comp continuous_map)
+                         (continuous_const.mul continuous_norm) },
+    { intro x,
+      rw map_coe f.uniform_continuous,
+      simp only [f.le_op_norm x, norm_coe] }
+  end }
+
+lemma normed_group_hom.completion_def (f : normed_group_hom G H) (x : completion G) :
+  f.completion x = completion.map f x := rfl
+
+lemma normed_group_hom.completion_coe_to_fun (f : normed_group_hom G H) :
+  (f.completion : completion G → completion H) = completion.map f :=
+by { ext x, exact normed_group_hom.completion_def f x }
+
+@[simp]
+lemma normed_group_hom.completion_coe (f : normed_group_hom G H) (g : G) : f.completion g = f g :=
+completion.map_coe f.uniform_continuous _
+
+@[simp]
+lemma normed_group_hom.completion_id : (normed_group_hom.id G).completion =
+  normed_group_hom.id (completion G) :=
+begin
+  ext x,
+  rw [normed_group_hom.completion_def, normed_group_hom.coe_id, completion.map_id],
+  refl
+end
+
+lemma normed_group_hom.completion_comp (f : normed_group_hom G H) (g : normed_group_hom H K) :
+  g.completion.comp f.completion = (g.comp f).completion :=
+begin
+  ext x,
+  rw [normed_group_hom.coe_comp, normed_group_hom.completion_def,
+    normed_group_hom.completion_coe_to_fun, normed_group_hom.completion_coe_to_fun,
+    completion.map_comp (normed_group_hom.uniform_continuous _)
+    (normed_group_hom.uniform_continuous _)],
+  refl
+end
+
+lemma normed_group_hom.completion_neg (f : normed_group_hom G H) :
```

Comment:
Does it make sense to define the group hom
```
(normed_group_hom G H) ->+ (normed_group_hom (completion G) (completion H))
```
because then you only have to prove the `completion_add` lemma.

After that `completion_{neg,sub,sum,zero}` will all have 1-line proofs.